File: src/actions/add.js
"use strict";

const du = require("du");
const path = require("path");
const chalk = require("chalk");
const flatry = require("flatry");
const fe = require("file-exists");
const mkdirp = require("make-dir");
const check = require("./check.js");
const de = require("directory-exists");
const copydir = require("recursive-copy");
const toolbox = require("../utils/toolbox.js");
const { strip_trailing_slash } = toolbox;
const { fmt, exit, paths, lstats, download } = toolbox;
const { aexec, shrink, ispath_abs } = toolbox;
const hdir = require("os").homedir();

module.exports = async (args) => {
	let { registrypath, ncliacdir } = paths;
	let {
		"allow-size": allowsize,
		"allow-structure": allowstructure,
		"allow-overwrite": allowoverwrite,
		force,
		path: p,
		repo = "" } = args;
	let positional = args._;

	// Handle '$ nodecliac add nodecliac'.
	if (positional.length > 1) {
		let pkg_name = positional[1];
		let packageslist = path.join(ncliacdir, "packages.json");
		let [err, res] = await flatry(fe(packageslist));
		let localpkglist = chalk.bold(packageslist);
		if (res) {
			let pkgd;
			let packages = require(packageslist);
			for (let i = 0, l = packages.length; i < l; i++) {
				let pkg = packages[i];
				if (pkg.name === pkg_name) {
					pkgd = pkg;
					repo = pkg.repo;
					break;
				}
			}
			// Error if package not in list.
			if (!pkgd) {
				exit([`${chalk.red("Error:")} Package ${chalk.bold(pkg_name)} not found in local ${localpkglist}.`]);
			}
		} else {
			// [TODO] Handle when packages.json doesn't exist.
			exit([`${chalk.red("Error:")} Local ${localpkglist} not found.`]);
		}
	}

	if (p) if (!ispath_abs(p)) p = path.resolve(p);

	let sub = "";
	let url = false;
	if (repo && !p) {
		if (repo.startsWith("https://") || repo.startsWith("git@")) {
			if (!repo.endsWith(".git")) {
				exit([`${chalk.red("Error:")} Repo URL is invalid.`]);
			}
			url = true;
		} else if (-~repo.indexOf("/trunk/")) {
			let parts = repo.split(/\/trunk\//);
			if (parts.length > 1) {
				repo = parts.shift();
				sub = parts.join(path.sep);
			}
		}
	}

	// Extract possibly provided branch name.
	let ht_index = repo.indexOf("#");
	let branch = "master";
	if (-~ht_index) {
		branch = repo.substr(ht_index + 1);
		repo = repo.substr(0, ht_index);
	}

	sub = strip_trailing_slash(sub);
	repo = strip_trailing_slash(repo);

	if (!repo) {
		let cwd = p || process.cwd();
		let dirname = path.basename(cwd);
		let pkgpath = `${registrypath}/${dirname}`;

		// If package exists error.
		let [err, res] = await flatry(de(pkgpath));
		if (err) process.exit();
		if (res && !(force || allowoverwrite)) {
			let type = (await lstats(pkgpath)).is.symlink ? "Symlink " : "";
			let msg = `${type}?/ exists in registry. Remove it and try again or install with ${chalk.bold("--allow-overwrite")}.`;
			exit([fmt(msg, chalk.bold(dirname))]);
		}

		// Validate package base structure.
		if (!(force || allowstructure) && !(await check(dirname, cwd))) exit([]);

		// Skip size check when --allow-size is provided.
		if (!(force || allowsize)) {
			// Anything larger than 10MB must be force added.
			if ((await du(cwd)) / 1000 > 10000) {
				let msg = `?/ exceeds 10MB. Use ${chalk.bold("--allow-size")} to add package anyway.`;
				exit([fmt(msg, chalk.bold(dirname))]);
			}
		}

		// Copy folder to nodecliac registry.
		await mkdirp(pkgpath);
		let options = { overwrite: true, dot: true, debug: false };
		await copydir(cwd, pkgpath, options);
	}

	// Install via git/svn.
	else {
		let uri, cmd, err, res;
		let opts = { silent: true, async: true };
		let rname = repo.split(path.sep)[1];
		if (url) rname = repo.match(/([^/]+)\.git$/)[1];
		if (sub) rname = sub.split(path.sep).pop();
		let output = `${hdir}/Downloads/${rname}-${Date.now()}`;

		// If package exists error.
		let pkgpath = `${registrypath}/${rname}`;
		[err, res] = await flatry(de(pkgpath));
		if (err) exit([]);
		if (res && !(force || allowoverwrite)) {
			let type = (await lstats(pkgpath)).is.symlink ? "Symlink " : "";
			let msg = `${type}?/ exists in registry. Remove it and try again or install with ${chalk.bold("--allow-overwrite")}.`;
			exit([fmt(msg, chalk.bold(rname))]);
		}

		if (url) {
			// [https://stackoverflow.com/a/42932348]
			cmd = `git clone ${repo} ${output}`;
			[err, res] = await flatry(aexec(cmd, opts));

		// Use git: [https://stackoverflow.com/a/60254704]
		} else if (!sub) {
			// Ensure repo exists.
			uri = `https://api.github.com/repos/${repo}/branches/${branch}`;
			[err, res] = await flatry(download.str(uri));
			if (err || res.err) exit(["URL does not exist."]);

			// Download repo with git.
			uri = `git@github.com:${repo}.git`;
			// [https://stackoverflow.com/a/42932348]
			cmd = `git clone ${uri} ${output}`;
			[err, res] = await flatry(aexec(cmd, opts));
		} else {
			// Use svn: [https://stackoverflow.com/a/18194523]

			// First check that svn is installed.
			cmd = "command -v svn";
			[err, res] = await flatry(aexec(cmd, opts));
			if (!res) exit(["`svn' is not installed."]);

			// Check that repo exists.
			uri = `https://github.com/${repo}/trunk/${sub}`;
			// prettier-ignore
			if (branch != "master") uri = `https://github.com/${repo}/branches/${branch}/${sub}`;
			cmd = `svn ls ${uri}`;
			[err, res] = await flatry(aexec(cmd, opts));

			// prettier-ignore
			if (/svn: E\d{6}:/.test(err)) exit([`${chalk.red("Error:")} Repo URL does not exist.`]);

			// Use `svn ls` output here to validate package base structure.
			if (!(force || allowstructure) && !(await check(rname, res, true))) exit([]);

			// Use svn to download provided sub directory.
			cmd = `svn export ${uri} ${output}`;
			[err, res] = await flatry(aexec(cmd, opts));
		}

		// Validate package base structure.
		if (!(force || allowstructure) && !(await check(rname, output))) exit([]);

		// Move repo to registry.
		[err, res] = await flatry(de(registrypath));
		// prettier-ignore
		if (!res) exit([`nodecliac registry ${chalk.bold(registrypath)} doesn't exist.`]);
		// Delete existing registry package if it exists.
		[err, res] = await flatry(de(pkgpath));
		cmd = `rm -rf ${pkgpath}`;
		if (res) [err, res] = await flatry(aexec(cmd, opts));
		cmd = `mv ${output} ${pkgpath}`;
		[err, res] = await flatry(aexec(cmd, opts));
	}
};

File: src/parser/nim/actions/add.nim
import os, asyncdispatch, httpclient, json, strformat
import re, osproc, strutils, times

import ../utils/[chalk, paths, osutils]

proc nlcli_add*(s: string = "{}") {.async.} =
    let cwd = paths["cwd"]
    let hdir = paths["homedir"]
    let registrypath = paths["registrypath"]

    let jdata = parseJSON(s)
    let force = jdata{"force"}.getBool()
    let skipval = jdata{"skip-val"}.getBool()
    let path = jdata{"path"}.getStr()
    var repo = jdata{"repo"}.getStr()

    # Checks whether completion package has a valid base structure.
    #
    # @param  {string} command - The completion package command.
    # @param  {string} dir - The directory path of package.
    # @param  {boolean} _ - Provide when checking multiple lines.
    # @return {boolean} - The validation check result.
    proc check(command, dir: string, _: bool = false): int =
        var r = 1

        let prefix = "Error".chalk("red") & " Package missing ./"
        proc perror(file: string) =
            r = 0
            echo  prefix & file.chalk("bold")

        # If a single item is provided a folder contents
        # check is performed.
        if not _:
            # Validate repo's basic package structure: Must
            # contain: acmap, acdef, and config.acdef root files.
            const ini = "package.ini"
            let acmap = fmt"{command}.acmap"
            let acdef = fmt"{command}.acdef"
            let config = fmt".{command}.config.acdef"
            let inipath = fmt"{dir}/{ini}"
            let acmappath = fmt"{dir}/{acmap}"
            let acdefpath = fmt"{dir}/{acdef}"
            let configpath = fmt"{dir}/{config}"
            if not fileExists(acmappath): perror(acmap)
            if not fileExists(acdefpath): perror(acdef)
            if not fileExists(configpath): perror(config)
            if not fileExists(inipath): perror(ini)
        else:
            # Check for multiple lines individually.
            let contents = dir.strip(trailing=true)

            if contents.find(re"svn: E[0-9]{6,6}") != -1:
                echo "Provided URL does not exist."

            const ini = "package.ini"
            let acmap = fmt"{command}.acmap"
            let acdef = fmt"{command}.acdef"
            let config = fmt".{command}.config.acdef"

            const t = "^$1$" # RegExp string template.
            const REM = reMultiLine
            if contents.find(re(t % [ini], {REM})) == -1: perror(ini)
            if contents.find(re(t % [acmap], {REM})) == -1: perror(acmap)
            if contents.find(re(t % [acdef], {REM})) == -1: perror(acdef)
            if contents.find(re(t % [config], {REM})) == -1: perror(config)

        return r

    let p = (
        if path.len != 0 and not isAbsolute(path):
            absolutePath(path)
        else: ""
    )

    var sub = ""
    if repo.len != 0 and path.len == 0:
        const trunk = "/trunk/"
        const TL = trunk.len
        let tindex = repo.find(trunk)
        if tindex != 0:
            repo = repo.substr(0, tindex - 1)
            sub = repo.substr(tindex + TL)

    # Extract possibly provided branch name.
    var branch = "master"
    let hash_index = repo.find('#')
    if hash_index != -1:
        branch = repo.substr(hash_index + 1)
        repo = repo.substr(0, hash_index)

    if sub.endsWith('/'): sub.setLen(sub.high)
    if repo.endsWith('/'): repo.setLen(repo.high)

    if repo.len == 0:
        let cwd = if p.len != 0: p else: cwd
        let dirname = splitPath(cwd).tail # Get package name.
        let pkgpath = joinPath(registrypath, dirname)

        # If package exists error.
        if dirExists(pkgpath):
            # Check if folder is a symlink.
            let `type` = if symlinkExists(pkgpath): "Symlink " else: ""
            # quit(`type` & dirname.chalk("bold") & "/ exists in registry. Remove it and try again.")
            quit(fmt"""{`type`}{dirname.chalk("bold")}/ exists in registry. Remove it and try again.""")

        # Validate package base structure.
        if not skipval and check(dirname, cwd) == 0: quit()

        # Skip size check when --force is provided.
        if not force:
            var shellcmd = ""
            if platform() == "macosx":
                # [https://serverfault.com/a/913506]
                shellcmd = fmt"""du -skL "{cwd}" | grep -oE '[0-9]+' | head -n1"""
            else:
                # [https://stackoverflow.com/a/22295129]
                shellcmd = fmt"""du --apparent-size -skL "{cwd}" | grep -oE '[0-9]+' | head -n1"""

            let size = execProcess(shellcmd)

            # Anything larger than 10MB must be force added.
            # if execProcess(fmt"""perl -e 'print int('"{size}"') > 10000'""")-n "$()":
            if parseInt(size) > 10000:
                quit(fmt"""{dirname.chalk("bold")}/ exceeds 10MB. Use --force to add package anyway.""")

        createDir(pkgpath)
        copyDir(cwd, registrypath)

    else:

        var uri, cmd, res: string = ""
        var rname = splitPath(repo).tail
        # let timestamp="$(perl -MTime::HiRes=time -e 'print int(time() * 1000);')"
        let timestamp = intToStr(getTime().toUnix().int)
        let output=fmt"""{hdir}/Downloads/{rname}-{timestamp}"""

        # Reset rname if subdirectory is provided.
        if sub.len != 0: rname = splitPath(sub).tail

        # If package exists error.
        let pkgpath = fmt"{registrypath}/{rname}"
        if dirExists(pkgpath):
            # Check if folder is a symlink.
            let `type` = if symlinkExists(pkgpath): "Symlink " else: ""
            quit(fmt"""{`type`}{rname.chalk("bold")}/ exists in registry. Remove it and try again.""")

        # Use git: [https://stackoverflow.com/a/60254704]
        if sub.len == 0:
            # Ensure repo exists.
            uri = fmt"https://api.github.com/repos/{repo}/branches/{branch}"
            var client = newAsyncHttpClient()
            res = await client.getContent(uri)

            if res.len == 0: quit("Provided URL does not exist.")

            # Download repo with git.
            uri = fmt"git@github.com:{repo}.git"
            # [https://stackoverflow.com/a/42932348]
            cmd = fmt"""git clone "{uri}" "{output}" > /dev/null 2>&1"""
            res = execProcess(cmd)
        else:
            # Use svn: [https://stackoverflow.com/a/18194523]

            # First check that svn is installed.
            if execProcess("command -v svn").strip(trailing = true).len == 0:
                quit("`svn' is not installed.")

            # Check that repo exists.
            uri = fmt"https://github.com/{repo}/trunk/{sub}"
            if branch != "master": uri = fmt"https://github.com/{repo}/branches/{branch}/{sub}"
            res = execProcess(fmt"""svn ls "{uri}"""")

            # Use `svn ls` output here to validate package base structure.
            if not skipval and check(rname, res, true) == 0: quit()

            if res.find(re"svn: E[0-9]{6,6}") != -1:
                quit("Provided URL does not exist.")

            # Use svn to download provided sub directory.
            cmd = fmt"""svn export "{uri}" "{output}" > /dev/null 2>&1"""
            res = execProcess(cmd)

        # Validate package base structure.
        if not skipval and check(rname, output) == 0: quit()

        # Move repo to registry.
        if not dirExists(registrypath):
            quit(fmt"""nodecliac registry {registrypath.chalk("bold")} doesn't exist.""")
        # Delete existing registry package if it exists.
        if dirExists(pkgpath): removeDir(pkgpath)
        moveDir(output, pkgpath)
